File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/account/infrastructure/notification/SesConfig.java
package com.example.accountservice.account.infrastructure.notification;

import com.example.accountservice.config.AwsProperties;
import java.net.URI;
import lombok.RequiredArgsConstructor;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.ses.SesClient;
import software.amazon.awssdk.services.ses.SesClientBuilder;

/**
 * AWS SES client bean configuration. The default credential provider chain (IMDS, etc.) has a large
 * delay before failing in a sandbox environment, so static credentials (StaticCredentialsProvider)
 * are always used explicitly.
 */
@Configuration
@RequiredArgsConstructor
public class SesConfig {

    private final AwsProperties awsProperties;

    @Bean
    public SesClient sesClient() {
        SesClientBuilder builder =
                SesClient.builder()
                        .region(Region.of(awsProperties.region()))
                        .credentialsProvider(
                                StaticCredentialsProvider.create(
                                        AwsBasicCredentials.create(
                                                awsProperties.accessKeyId(),
                                                awsProperties.secretAccessKey())));

        if (awsProperties.endpointUrl() != null && !awsProperties.endpointUrl().isBlank()) {
            builder.endpointOverride(URI.create(awsProperties.endpointUrl()));
        }

        return builder.build();
    }
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/notification/infrastructure/SesConfig.kt
package com.example.accountservice.notification.infrastructure

import com.example.accountservice.config.AwsProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider
import software.amazon.awssdk.regions.Region
import software.amazon.awssdk.services.ses.SesClient
import java.net.URI

@Configuration
class SesConfig(
    private val awsProperties: AwsProperties,
) {
    @Bean
    fun sesClient(): SesClient {
        // Always set credentials explicitly. Relying on DefaultCredentialsProvider (IMDS, etc.) can slow
        // down the response during the lookup process when credentials aren't available.
        val builder =
            SesClient
                .builder()
                .region(Region.of(awsProperties.region))
                .credentialsProvider(
                    StaticCredentialsProvider.create(
                        AwsBasicCredentials.create(awsProperties.accessKeyId, awsProperties.secretAccessKey),
                    ),
                )

        if (awsProperties.endpointUrl.isNotBlank()) {
            builder.endpointOverride(URI.create(awsProperties.endpointUrl))
        }

        return builder.build()
    }
}
